main.go

```diff
@@ -219,3 +237,137 @@ func getCacheFunc() (cache.NewCacheFunc, error) {
 		SelectorsByObject: selectors,
 	}), nil
 }
+
+// restConfigForAPIExport returns a *rest.Config properly configured to communicate with the endpoint for the
+// APIExport's virtual workspace.
+func restConfigForAPIExport(ctx context.Context, cfg *rest.Config, apiExportName string) (*rest.Config, error) {
+	scheme := runtime.NewScheme()
+	if err := apisv1alpha1.AddToScheme(scheme); err != nil {
+		return nil, fmt.Errorf("error adding apis.kcp.dev/v1alpha1 to scheme: %w", err)
+	}
+
+	apiExportClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	if err != nil {
+		return nil, fmt.Errorf("error creating APIExport client: %w", err)
+	}
+
+	var apiExport apisv1alpha1.APIExport
+
+	if apiExportName != "" {
+		if err := apiExportClient.Get(ctx, types.NamespacedName{Name: apiExportName}, &apiExport); err != nil {
+			return nil, fmt.Errorf("error getting APIExport %q: %w", apiExportName, err)
+		}
+	} else {
+		setupLog.Info("api-export-name is empty - listing")
+		exports := &apisv1alpha1.APIExportList{}
+		if err := apiExportClient.List(ctx, exports); err != nil {
+			return nil, fmt.Errorf("error listing APIExports: %w", err)
+		}
+		if len(exports.Items) == 0 {
+			return nil, fmt.Errorf("no APIExport found")
+		}
+		if len(exports.Items) > 1 {
+			return nil, fmt.Errorf("more than one APIExport found")
+		}
+		apiExport = exports.Items[0]
+	}
+
+	if len(apiExport.Status.VirtualWorkspaces) < 1 {
+		return nil, fmt.Errorf("APIExport %q status.virtualWorkspaces is empty", apiExportName)
+	}
+
+	cfg = rest.CopyConfig(cfg)
+	cfg.Host = apiExport.Status.VirtualWorkspaces[0].URL
+
+	return cfg, nil
+}
+
+func kcpAPIsGroupPresent(restConfig *rest.Config) bool {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		setupLog.Error(err, "failed to create discovery client")
+		os.Exit(1)
+	}
+	apiGroupList, err := discoveryClient.ServerGroups()
+	if err != nil {
+		setupLog.Error(err, "failed to get server groups")
+		os.Exit(1)
+	}
+
+	for _, group := range apiGroupList.Groups {
+		if group.Name == apisv1alpha1.SchemeGroupVersion.Group {
+			for _, version := range group.Versions {
+				if version.Version == apisv1alpha1.SchemeGroupVersion.Version {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func ensureRequiredAPIGroupsAndResourcesExist(restConfig *rest.Config) {
```

Comment:
I will probably borrow this for jvm-build-service :-)

Reply:
Why not let the container gets into CrashLoopBackOff if an API is not available?